pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/state/StateContext.java

```diff
@@ -18,15 +18,61 @@
  */
 package org.apache.pulsar.functions.instance.state;
 
+import java.nio.ByteBuffer;
 import java.util.concurrent.CompletableFuture;
 
 /**
  * A state context per function.
  */
 public interface StateContext {
 
+    /**
+     * Increment the given <i>key</i> by the given <i>amount</i>.
+     *
+     * @param key key to increment
+     * @param amount the amount incremented
+     */
     void incr(String key, long amount);
 
+    /**
+     * Update the given <i>key</i> to the provide <i>value</i>.
+     *
+     * <p>NOTE: the put operation might or might not be applied directly to the global state until
+     * the state is flushed via {@link #flush()} at the completion of function execution.
+     *
+     * <p>The behavior of `PUT` is non-deterministic, if two function instances attempt to update
+     * same key around the same time, there is no guarantee which update will be the final result.
+     * That says, if you attempt to get amount via {@link #getAmount(String)}, increment the amount
+     * based on the function computation logic, and update the computed amount back. one update will
+     * overwrite the other update. For this case, you are encouraged to use {@link #incr(String, long)}
+     * instead.
+     *
+     * @param key key to update.
+     * @param value value to update
+     */
+    void put(String key, ByteBuffer value);
+
+    /**
+     * Get the value of a given <i>key</i>.
+     *
+     * @param key key to retrieve
+     * @return a completable future representing the retrieve result.
+     */
+    CompletableFuture<ByteBuffer> getValue(String key);
```

Comment:
IMO either both put/get shd return future. Or none of them should.
I would vote for non future based. It keeps api simple. If we need faster/complicated future based api, we can add getAsync/putAsync later?

Reply:
just a reminder, getValue can be very slow. it is not a local state, it is a global state. it involves network I/O and potential disk I/O. do we think we want to do sync calls only?